[webkitscmpy] Read and write BitBucket comments

JonWBedard opened the pull request:
#### 202bb7baa3ad3688b5fb23f74782cd7ee52ced18
<pre>
[webkitscmpy] Read and write BitBucket comments
<a href="https://bugs.webkit.org/show_bug.cgi?id=273007">https://bugs.webkit.org/show_bug.cgi?id=273007</a>
<a href="https://rdar.apple.com/126769869">rdar://126769869</a>

Reviewed by Elliott Williams.

Provide Python functions to both write and read in-line PR comments.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/bitbucket.py:
(BitBucket.request): Mock endpoints to write and read PR comments.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
(BitBucket.PRGenerator.comment): Allow comments to define an anchor.
(BitBucket.PRGenerator.comments): Skip actions which have &apos;commentAnchor&apos;.
(BitBucket.PRGenerator._position_converters): Provide dictionaries to convert from
relative to absolute diff positions and visa versa.
(BitBucket.PRGenerator._diff_comments): Return a list of comments anchored to a relative
diff position.
(BitBucket.PRGenerator.review): Add &apos;diff_comments&apos; argument.
(BitBucket.PRGenerator.diff): Added.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:

Canonical link: <a href="https://commits.webkit.org/278406@main">https://commits.webkit.org/278406@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6cd28709dacfcd14da4c92ef36e366d9484883f2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29760 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53723 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/1154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52767 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36006 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/804 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/41161 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52563 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/27421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/43446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/22266 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/50315 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/24831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8842 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/46811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55312 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/25562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48567 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/50485 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/43604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47610 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/27687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7294 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/26555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
```diff
@@ -322,40 +339,171 @@ def comments(self, pull_request):
 
+        @classmethod
+        def _position_converters(cls, json_diff):
+            abosolute_to_relative = dict(
```
typo: s/abosolute/absolute

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
```diff
@@ -322,40 +339,171 @@ def comments(self, pull_request):
+        @classmethod
+        def _position_converters(cls, json_diff):
+            abosolute_to_relative = dict(
+                source=defaultdict(lambda: defaultdict(None)),
```
is `defaultdict(lambda: defaultdict(None))` any different from a `defaultdict(dict)`?

JonWBedard replied:
It means you don't need to do `.get` to get a `None`, `[]` will also work. I made this a bit more explicit, though, because `None` often has magical properties in Python, and it's not intended to here.

emw-apple reviewed (approved):
I will not pretend to fully understand the diff line counting arithmetic, but this makes sense and seems well-structured.

JonWBedard committed:
[webkitscmpy] Read and write BitBucket comments
https://bugs.webkit.org/show_bug.cgi?id=273007
rdar://126769869

Reviewed by Elliott Williams.

Provide Python functions to both write and read in-line PR comments.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/bitbucket.py:
(BitBucket.request): Mock endpoints to write and read PR comments.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/remote/bitbucket.py:
(BitBucket.PRGenerator.comment): Allow comments to define an anchor.
(BitBucket.PRGenerator.comments): Skip actions which have 'commentAnchor'.
(BitBucket.PRGenerator._position_converters): Provide dictionaries to convert from
relative to absolute diff positions and visa versa.
(BitBucket.PRGenerator._diff_comments): Return a list of comments anchored to a relative
diff position.
(BitBucket.PRGenerator.review): Add 'diff_comments' argument.
(BitBucket.PRGenerator.diff): Added.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:

Canonical link: https://commits.webkit.org/278406@main
